[PLATFORM-454]: [Auth0Ex] Add default middleware for Absinthe

cpiemontese opened the pull request:
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-454

This PR adds an Absinthe plug and middlerware developed by @enerdgumen for BBurago.

I am unsure about a few things:

1. should this plug/middleware live in this repo or in its own repo?
2. should Absinthe be included as an optional or test-only dependency?

mbusi commented on lib/prima_auth0_ex/plug/absinthe.ex:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  defmodule RequirePermission do
```
why isn't hits module in its own file?

enerdgumen replied:
It was just for cohesion, since these modules depend on each other and are both closely coupled with Absinthe, so in Bburago I put them close in the same file.

If you want to split them I suggest to move them in a namespace specific for absinthe (for instance `PrimaAuth0Ex.Absinthe.CreateSecurityContext/RequirePermission`).

mbusi commented on lib/prima_auth0_ex/plug/absinthe.ex:
```diff
@@ -0,0 +1,81 @@
+          resolution
+
+        dry_run and permissions != nil ->
+          Logger.warn("Received invalid token", required_permissions: [required_permission])
```
why a list if the argument is always a single permission?

enerdgumen replied:
Because for Bburago it's enough, but here it would be better to support a list of permissions.

mbusi replied:
the argument should be a list then, I guess :D

cpiemontese replied:
Yeah I'll change it

mbusi commented on lib/prima_auth0_ex/plug/absinthe.ex:
```diff
@@ -0,0 +1,81 @@
+          %{context: %Context{dry_run: dry_run, permissions: permissions}} = resolution,
+          required_permission
+        ) do
+      cond do
```
instead of a `cond` we could split this into two different functions, one to check for permissions, the other one to act on the resolution (accepting `dry_run` as argument)

mbusi reviewed (commented):
I think having absinthe as an optional dependency makes sense, we could test what happens in a project including auth0_ex but NOT absinthe :D

mbusi commented:
Forgot to answer to
> 1. should this plug/middleware live in this repo or in its own repo?

I'd keep it here (if it's not too annoying for dependencies management)

cpiemontese commented:
I refactored a bit

- absinthe version requirement is now lowered to 1.6
- I split the absinthe modules into `plug` and `middleware`
- I tried to refactor the `RequirePermission` module

enerdgumen commented on mix.exs:
```diff
@@ -47,6 +47,8 @@ defmodule PrimaAuth0Ex.MixProject do
   defp dev_deps do
     [
+      {:absinthe, "~> 1.6", optional: true},
+      {:absinthe_plug, "~> 1.5", optional: true},
```
These are not `dev_deps`.

I would mark even `plug` as an optional dependency (probably also `redix`, since the caching is optional, but in such a case we should change the default to false).

cpiemontese replied:
You're right, I fixed the dependencies!

`redix` could be optional too, but I would leave it for a future PR

enerdgumen commented on lib/prima_auth0_ex/middleware/require_permission.ex:
```diff
@@ -0,0 +1,42 @@
+  def call(
+        %{context: %Context{dry_run: dry_run, permissions: permissions}} = resolution,
+        required_permission
+      ) do
```
Is all this really clearer than a trivial set of [conditions](https://github.com/primait/bburago/blob/master/lib/bburago/graphql/authorization.ex#L66)? 🙂

mbusi replied:
I guess this could be more readable with an hybrid approach, something like
```elixir
if has_required_permission?(permissions, required_permission) do
  resolution
 else
  resolve(dry_run, resolution)
end
```

mbusi commented on lib/prima_auth0_ex/middleware/require_permission.ex:
```diff
@@ -0,0 +1,42 @@
+      # On dry-run, log a warning if token doesn't have required permission
+      # but permissions was not nil (i.e. it was supposed to work)
+      if not has_required_permissions? and permissions != nil and dry_run do
+        Logger.warn("Received invalid token", required_permission: required_permission)
```
shouldn't this simply go into the resolve function (in the corresponding match?)

mbusi commented on lib/prima_auth0_ex/middleware/require_permission.ex:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  @spec has_required_permission?(permissions :: [any()] | nil, required_permission :: any()) :: boolean()
+  defp has_required_permission?(nil, _), do: false
```
tbh the case for `nil` required_permissions shouldn't even be allowed...

mbusi commented on lib/prima_auth0_ex/absinthe/require_permissions.ex:
```diff
@@ -0,0 +1,39 @@
+  defp has_required_permissions?(nil = _permissions, _required_permissions), do: false
+
+  defp has_required_permissions?(permissions, required_permissions),
+    do: Enum.all?(required_permissions, fn required_permission -> required_permission in permissions end)
```
```suggestion
    do: Enum.all?(required_permissions, &Enum.member?(required_permissions, &1))
```
Dunno if I like it more this way :laughing:

enerdgumen commented on test/integration/absinthe_test.exs:
```diff
@@ -0,0 +1,92 @@
+defmodule PrimaAuth0Ex.Plug.AbsintheTest do
+  # , async: true
```
dead code

enerdgumen reviewed (approved):
❤️
